src/room/room-prototype.ts

```diff
@@ -63,18 +64,93 @@ const findNextEnergySource = function(creep: Creep) {
             }
         }
     }
-}
+};
+
+const getNumberOfMiningSpaces = function() {
+    if (!this.memory['sources'] || !this.memory['sources']['spaces']) {
+        this.findNumberOfSourcesAndSpaces();
+    }
+    return this.memory['sources']['spaces'];
+};
+
+const getNumberOfSources = function() {
+    if (!this.memory['sources'] || !this.memory['sources']['count']) {
+        this.findNumberOfSourcesAndSpaces();
+    }
+    return this.memory['sources']['count'];
+};
+
+const findNumberOfSourcesAndSpaces = function() {
+    let numberSources = 0;
+    let numberSpaces = 0;
+    _.forEach(this.find(FIND_SOURCES), (source: Source) => {
+        numberSources++;
+        const availablePositions = {};
+        for (let x = source.pos.x-1; x < source.pos.x + 1; x++) {
+            for (let y = source.pos.y-1; y < source.pos.y + 1; y++) {
+                availablePositions[x + ":" + y] = true;
+            }
+        }
+        _.forEach(this.lookAtArea(source.pos.y-1, source.pos.x-1, source.pos.y+1, source.pos.x+1, true),
```

Comment:
`lookAtArea` returns multiple LookAtResultWithPos for the same coord. You could have a swamp, creep, road, and rampart all on the same spot and it would return 4 results for that coord. That's why I can't do that.